pkg/gc/gc_state_manager.go

```diff
@@ -0,0 +1,567 @@
+		var err1 error
+		oldTxnSafePoint, err1 = m.gcMetaStorage.LoadTxnSafePoint(keyspaceID)
+		if err1 != nil {
+			return err1
+		}
+
+		if target < oldTxnSafePoint {
+			return errs.ErrDecreasingTxnSafePoint.GenWithStackByArgs(oldTxnSafePoint, target)
+		}
+
+		barriers, err1 := m.gcMetaStorage.LoadAllGCBarriers(keyspaceID)
+		if err1 != nil {
+			return err1
+		}
+
+		for _, barrier := range barriers {
+			if keyspaceID == constant.NullKeyspaceID && barrier.BarrierID == keypath.GCWorkerServiceSafePointID {
+				downgradeCompatibleMode = true
+				continue
+			}
+
+			if barrier.IsExpired(now) {
+				// Perform lazy delete to the expired GC barriers.
+				// WARNING: It might look like a reasonable optimization idea to perform the lazy-deletion in a lower
+				// frequency (instead of everytime checking it). However, it's UNSAFE considering the possibility of
+				// system clock drifts and PD leader changes, in which case an expired GC barrier may be back to
+				// not-expired state again. Once we regard a GC barrier as expired, it must be expired *strictly*,
+				// otherwise it may break the constraint that GC barriers must block the txn safe point from being
+				// advanced over them.
+				err1 = wb.DeleteGCBarrier(keyspaceID, barrier.BarrierID)
+				if err1 != nil {
+					return err1
+				}
+				// Do not block GC with expired barriers.
+				continue
+			}
+
+			if barrier.BarrierTS < minBlocker {
+				minBlocker = barrier.BarrierTS
+				blockingBarrier = barrier
+			}
+		}
+
+		// Compatible with old TiDB nodes that use TiDBMinStartTS to block GC.
+		ownerKey, minStartTS, err1 := m.gcMetaStorage.CompatibleLoadTiDBMinStartTS(keyspaceID)
+		if err1 != nil {
+			return err1
+		}
+
+		if minStartTS != 0 && len(ownerKey) != 0 && minStartTS < minBlocker {
+			// Note that txn safe point is defined inclusive: snapshots that exactly equals to the txn safe point are
+			// considered valid.
+			minBlocker = minStartTS
+			blockingBarrier = nil
+			blockingMinStartTSOwner = &ownerKey
+		}
+
+		// Txn safe point never decreases.
+		newTxnSafePoint = max(oldTxnSafePoint, minBlocker)
+
+		if downgradeCompatibleMode {
+			err1 = wb.SetGCBarrier(keyspaceID, endpoint.NewGCBarrier(keypath.GCWorkerServiceSafePointID, newTxnSafePoint, nil))
+			if err1 != nil {
+				return err1
+			}
+		}
+		return wb.SetTxnSafePoint(keyspaceID, newTxnSafePoint)
+	})
+	if err != nil {
+		return AdvanceTxnSafePointResult{}, err
+	}
+
+	blockerDesc := ""
+	if blockingBarrier != nil {
+		blockerDesc = blockingBarrier.String()
+	} else if blockingMinStartTSOwner != nil {
+		blockerDesc = fmt.Sprintf("TiDBMinStartTS { Key: %+q, MinStartTS: %d }", *blockingMinStartTSOwner, newTxnSafePoint)
+	}
+
+	if newTxnSafePoint != target {
+		if blockingBarrier == nil && blockingMinStartTSOwner == nil {
+			panic("unreachable")
+		}
+		if newTxnSafePoint == minBlocker {
+			log.Info("txn safe point advancement is being blocked",
```

Comment:
Why we not print log once `minBlocker` update? Such a log is more accurate and clear

Reply:
It's possible that `minBlocker` is updated multiple times when iterating over all potential blockers, producing too much log. We only care about the final result here.